### PR TITLE
(968a) Add HMPPS Manage Users API

### PR DESCRIPTION
**Pull request by jsrobertson**
## Context

We need to display the actual names of the user who makes a referral, not just their username.  To do this it means making use of https://github.com/ministryofjustice/hmpps-manage-users-api.

https://trello.com/c/K6mwNUHH/968-fetch-name-of-referrer-rather-than-username-for-programme-history-and-cya-page

## Changes in this PR
This PR is for integrating the HMPPS Manage Users API and in the first instance, replacing the existing `/user/me` endpoint from HMPPS Auth with the new `/users/me` and `/users/${username}` from HMPPS Manage Users API.

This also meant making some changes to some of the User types as the response schema in the new api is slightly different.

## Release checklist

[Release process documentation](../doc/how-to/perform-a-release.md)

As part of our continuous deployment strategy we must ensure that this work is
ready to be released once merged.

### Pre-merge

- [ ] There are changes required to the Accredited Programmes API for this change to work...
  - [ ] ... and they have been released to production already

### Post-merge

- [ ] This adds, extends or meaningfully modifies a feature...
  - [ ] ... and I have written or updated an end-to-end test for the happy path in the [Accredited Programmes E2E repo](https://github.com/ministryofjustice/hmpps-accredited-programmes-e2e)
- [ ] This makes new expectations of the API...
  - [ ] ... and I have notified the API developer(s) of changes to the contract tests (Pact), or the API is already compliant
- [ ] [Manually approve](../doc/how-to/perform-a-release.md#releasing-to-the-preprod-environment) release to preprod
- [ ] [Manually approve](../doc/how-to/perform-a-release.md#releasing-to-the-production-environment) release to prod

### CODE REVIEW

File: server/services/userService.ts [hunk 1]
```diff
@@ -1,19 +1,24 @@
...
-    private readonly hmppsAuthClientBuilder: RestClientBuilderWithoutToken<HmppsAuthClient>,
+    private readonly hmppsManageUsersClientBuilder: RestClientBuilder<HmppsManageUsersClient>,
     private readonly caseloadClientBuilder: RestClientBuilder<CaseloadClient>,
   ) {}
 
-  async getUser(token: Express.User['token']): Promise<UserDetails> {
-    const hmppsAuthClient = this.hmppsAuthClientBuilder()
-    const user = await hmppsAuthClient.getUser(token)
-    const caseloads = await this.getCaseloads(token)
+  async getUserWithDetails(token: Express.User['token']): Promise<UserDetails> {
```
Comment (jsrobertson) [inline server/services/userService.ts:13]: Wondering whether this should be called something else.  `getCurrentUserWithDetails` or just `getCurrentUser`?

File: server/data/hmppsManageUsersClient.test.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+      fakeManangeUsersApiUrl.get('/users/me').matchHeader('authorization', `Bearer ${token}`).reply(200, response)
+
+      const output = await hmppsManageUsersClient.getCurrentUsername()
+      expect(output).toEqual(response)
+    })
+  })
+
+  describe('getUserFromUsername', () => {
+    it('should return data from api', async () => {
```
Comment (Gweaton) [inline server/data/hmppsManageUsersClient.test.ts:41]: I know it's kind of implicit here, but how about: "should return the user from the API"?

File: server/data/hmppsManageUsersClient.test.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    if (!nock.isDone()) {
+      nock.cleanAll()
+      throw new Error('Not all nock interceptors were used!')
+    }
+    nock.abortPendingRequests()
+    nock.cleanAll()
+  })
+
+  describe('getCurrentUsername', () => {
+    it('should return data from api', async () => {
```
Comment (Gweaton) [inline server/data/hmppsManageUsersClient.test.ts:30]: "should return the logged-in user's username from the api"?

File: server/data/hmppsManageUsersClient.ts [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+  constructor(token: Express.User['token']) {
+    this.restClient = new RestClient('HMPPS Manage Users Client', config.apis.hmppsManageUsers, token)
+  }
+
+  getCurrentUsername(): Promise<Pick<User, 'username'>> {
+    logger.info(`Getting username: calling HMPPS Manage Users`)
+    return this.restClient.get({ path: '/users/me' }) as Promise<Pick<User, 'username'>>
+  }
+
+  getUserFromUsername(username: string): Promise<User> {
```
Comment (jsrobertson) [inline server/data/hmppsManageUsersClient.ts:18]: or `findUserFromUsername`

File: server/services/userService.ts [hunk 5]
```diff
@@ -1,19 +1,24 @@
...
+    const hmppsManageUsersClient = this.hmppsManageUsersClientBuilder(token)
+    const { username } = await hmppsManageUsersClient.getCurrentUsername()
+
+    const [user, caseloads] = await Promise.all([
+      hmppsManageUsersClient.getUserFromUsername(username),
+      this.getCaseloads(token),
+    ])
```
Comment (yndajas) [inline server/services/userService.ts:20]: Is this format preferable to just having two `await` lines (like before)?

File: server/services/userService.ts (hunk 1 above)
Comment (Gweaton) [reply server/services/userService.ts:13]: I think `getCurrentUserWithDetails` works well

**Review by Gweaton — APPROVED** [review]
Looks good, well done for getting this all figured out. Just have a couple of small naming comments that would be good to take care of before merging.

**Review by yndajas — COMMENTED** [review]
Looks good! Just a few small things

File: server/services/userService.ts (hunk 5 above)
Comment (jsrobertson) [reply server/services/userService.ts:20]: Yeah I hadn't done it anywhere else up to now (although not sure if there's been a reason to) but I tend to prefer it as it runs them concurrently.

Comment (yndajas) [reply server/services/userService.ts:20]: Ooh nice! Might be good to start doing that then

File: docker-compose.yml [hunk 6]
```diff
@@ -52,6 +54,77 @@ services:
     ports:
       - '9090:8080'
 
+  auth-db:
```
[the remark this answers is not in the document]
Comment (jsrobertson) [reply docker-compose.yml:57]: Good spot, I was meant to add a `depends `hmpps-auth`  too.  Will rename to `hmpps-auth-db`

File: docker-compose.yml [hunk 7]
```diff
@@ -52,6 +54,77 @@ services:
...
+      - EXTERNAL_USERS_ENDPOINT_URL=http://hmpps-external-users-api:8098
+      - AUTHORIZATION_SERVER_TOKEN_ENDPOINT_URL=http://hmpps-auth:8080/auth/oauth/token
+      - NOMIS_ENDPOINT_URL=http://nomis-user-roles-api:8097
+    healthcheck:
+      test: [ "CMD", "curl", "-f", "http://localhost:8096/health" ]
+    image: quay.io/hmpps/hmpps-manage-users-api:latest
+    networks:
+      - hmpps
+    ports:
+      - "8096:8096"
```
[the remark this answers is not in the document]
Comment (jsrobertson) [reply docker-compose.yml:91]: No real reason other than it was what I'd seen some of the other projects too.  Will try and follow that convention where I can though, cheers!

File: server/data/hmppsManageUsersClient.test.ts [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+jest.mock('./tokenStore')
+
+const token = 'token-1'
+
+describe('HmppsManageUsersClient', () => {
+  let fakeManangeUsersApiUrl: nock.Scope
+  let hmppsManageUsersClient: HmppsManageUsersClient
+
+  beforeEach(() => {
+    fakeManangeUsersApiUrl = nock(config.apis.hmppsManageUsers.url)
```
[the remark this answers is not in the document]
Comment (jsrobertson) [reply server/data/hmppsManageUsersClient.test.ts:16]: 🙃

File: helm_deploy/values-dev.yaml [hunk 9]
```diff
@@ -13,6 +13,7 @@ generic-service:
     ACCREDITED_PROGRAMMES_API_URL: 'https://accredited-programmes-api-dev.hmpps.service.justice.gov.uk'
     INGRESS_URL: 'https://accredited-programmes-dev.hmpps.service.justice.gov.uk'
     HMPPS_AUTH_URL: 'https://sign-in-dev.hmpps.service.justice.gov.uk/auth'
+    HMPPS_MANAGE_USERS_API_URL: "https://manage-users-api-dev.hmpps.service.justice.gov.uk"
```
[the remark this answers is not in the document]
Comment (jsrobertson) [reply helm_deploy/values-dev.yaml:16]: Ah good spot I have prettier but hadn't ran format document for any YAML files yet so it was configured.  Sorted now.

File: server/data/hmppsManageUsersClient.ts (hunk 4 above)
Comment (jsrobertson) [reply server/data/hmppsManageUsersClient.ts]: I've left it as `getUserFromUsername` for now.  We have `getCurrentUsername` in there too but I'm not sure that works with `find`, so to have a mix of `find` and `get` feels a bit confusing.

File: docker-compose.yml [hunk 10]
```diff
@@ -52,6 +56,77 @@ services:
...
+      - EXTERNAL_USERS_ENDPOINT_URL=http://hmpps-external-users-api:8080
+      - AUTHORIZATION_SERVER_TOKEN_ENDPOINT_URL=http://hmpps-auth:8080/auth/oauth/token
+      - NOMIS_ENDPOINT_URL=http://nomis-user-roles-api:8080
+    healthcheck:
+      test: ['CMD', 'curl', '-f', 'http://localhost:8080/health']
+    image: quay.io/hmpps/hmpps-manage-users-api:latest
+    networks:
+      - hmpps
+    ports:
+      - '9095:8080'
```
Comment (jsrobertson) [inline docker-compose.yml:112]: `9093` and `9094` already referred to in config.ts

**Review by yndajas — APPROVED** [review]

**Pull request merged**